docs/core/knowledge-bases.rst

```diff
@@ -0,0 +1,416 @@
+You can customize your ``InMemoryKnowledgeBase`` by overwriting the following functions:
+
+- ``get_key_attribute_of_object``: To keep track of what object the user was talking about last, we store the value
+  of the key attribute in a specific slot. Every object should have a key attribute that is unique, i.e.
+  similar to the primary key in a relation database. By default the name of the key attribute for every object type
+  is set to "id". You can overwrite the name of the key attribute for a specific object type by calling
+  ``set_key_attribute_of_object()``.
+- ``get_representation_function_of_object``: Let's focus on the following restaurant:
+
+  .. code-block:: json
+
+      {
+          "id": 0,
+          "name": "Donath",
+          "cuisine": "Italian",
+          "outside-seating": true,
+          "price-range": "mid-range"
+      }
+
+  When the user is asking to list any Italian restaurant, you don't want to confront the user with all details of that
+  restaurant. You want to provide a meaningful name that identifies the restaurant. Most likely you would use
+  just the name of the restaurant to speak about it.
+  Thus, the function ``get_representation_function_of_object`` returns a lambda function that maps, for example, the
+  above restaurant object to its name.
+
+  .. code-block:: python
+
+      lambda obj: obj["name"]
+
+  This function is used whenever the bot is talking about a specific object, so that the user is given a meaningful
+  name and knows what exactly the bot is talking about.
+  By default the lambda function is set to ``lambda obj: obj["name"]``. So, it returns the value of the attribute
+  "name" of the object. If your object does not have an attribute "name", or the "name" of an object might be
+  ambiguous, you should set a new lambda function for that object type by calling
+  ``set_representation_function_of_object()``.
+- ``set_ordinal_mention_mapping``: The ordinal mention mapping is needed to resolve an ordinal mention to an object
+  in a list. For example, if the bot listed a few restaurants in Berlin, and the user then asked "Does the second one
+  have outside seating?", you need to resolve "second one" to the correct object the bot listed before. Per
+  default the ordinal mention mapping looks like this:
+
+  .. code-block:: python
+
+      {
+          "1": lambda l: l[0],
+          "2": lambda l: l[1],
+          "3": lambda l: l[2],
+          "4": lambda l: l[3],
+          "5": lambda l: l[4],
+          "6": lambda l: l[5],
+          "7": lambda l: l[6],
+          "8": lambda l: l[7],
+          "9": lambda l: l[8],
+          "10": lambda l: l[9],
+          "ANY": lambda l: random.choice(list),
+          "LAST": lambda l: l[-1],
+      }
+
+  You can overwrite it by calling the function ``set_ordinal_mention_mapping``.
+  If you want to learn more about the usage of the mapping, go to section :ref:`resolve_mentions`.
+
+Defining the NLU Data
```

Comment:
Let's start this section with a brief summary:

- we are going to introduce a new intent, ``query_knowledge_base``
- we are going to annotate `mention` entities so that our model detects indirect mentions of entities like 'the first one'
- we will use synonyms extensively